src/libsyntax/tokenstream.rs

```diff
@@ -417,112 +497,185 @@ fn covering_span(trees: &[TokenTree]) -> Span {
+            sp: new_span,
         }
-
     }
 
-    /// Convert a vector of TokenTrees into a parentheses-delimited TokenStream.
-    pub fn as_paren_delimited_stream(tts: Vec<TokenTree>) -> TokenStream {
-        let new_sp = covering_span(&tts);
-
-        let new_delim = Rc::new(Delimited {
-            delim: token::DelimToken::Paren,
-            open_span: DUMMY_SP,
-            tts: tts,
-            close_span: DUMMY_SP,
-        });
-
-        TokenStream::from_tts(vec![TokenTree::Delimited(new_sp, new_delim)])
+    /// Indicate if the TokenStream is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
     }
 
-    /// Convert an interned string into a one-element TokenStream.
-    pub fn from_interned_string_as_ident(s: InternedString) -> TokenStream {
-        TokenStream::from_tts(vec![TokenTree::Token(DUMMY_SP,
-                                                    Token::Ident(token::str_to_ident(&s[..])))])
+    /// Return a TokenSlice's length.
+    pub fn len(&self) -> usize {
+        match *self {
+            TokenStream::Empty => 0,
+            TokenStream::Leaf { ref tts, .. } => tts.len(),
+            TokenStream::SubLeaf { len, .. } => len,
+            TokenStream::Node { len, .. } => len,
+        }
     }
-}
 
-/// TokenSlices are 'views' of `TokenStream's; they fit the same role as `str`s do for
-/// `String`s. In general, most TokenStream manipulations will be refocusing their internal
-/// contents by taking a TokenSlice and then using indexing and the provided operators.
-#[derive(PartialEq, Eq, Debug)]
-pub struct TokenSlice([TokenTree]);
-
-impl ops::Deref for TokenStream {
-    type Target = TokenSlice;
+    /// Convert a TokenStream into a vector of borrowed TokenTrees.
+    pub fn to_vec(&self) -> Vec<&TokenTree> {
+        match *self {
+            TokenStream::Empty => Vec::new(),
+            TokenStream::Leaf { ref tts, .. } => tts.iter().collect(),
+            TokenStream::SubLeaf { ref tts, offset, len, .. } => {
+                tts[offset..len - 1].iter().collect()
+            }
+            TokenStream::Node { ref left, ref right, .. } => {
+                let mut v1 = TokenStream::to_vec(left);
+                let mut v2 = TokenStream::to_vec(right);
+                v1.append(&mut v2);
+                v1
+            }
+        }
+    }
 
-    fn deref(&self) -> &TokenSlice {
-        let tts: &[TokenTree] = &*self.tts;
-        unsafe { mem::transmute(tts) }
+    /// Convert a TokenStream into a vector of TokenTrees (by cloning the TokenTrees).
+    /// (This operation is expensive.)
+    pub fn to_tts(&self) -> Vec<TokenTree> {
+        TokenStream::to_vec(self).into_iter().cloned().collect::<Vec<TokenTree>>()
     }
-}
 
-impl TokenSlice {
-    /// Convert a borrowed TokenTree slice into a borrowed TokenSlice.
-    fn from_tts(tts: &[TokenTree]) -> &TokenSlice {
-        unsafe { mem::transmute(tts) }
+    /// Manually change a TokenStream's span.
```

Comment:
This returns, rather than changes the span, no?

I'm not sure how I feel about Empty returning DUMMY_SP, perhaps this should return an Option<Span>? Or maybe Empty TokenStreams should have a span, just a zero-length one? Not sure if that would be too annoying.